pkg/route/secret/secret_monitor.go

```diff
@@ -0,0 +1,215 @@
+	//
+	RemoveSecretEventHandler(SecretEventHandlerRegistration) error
+	//
+	GetSecret(SecretEventHandlerRegistration) (*v1.Secret, error)
+}
+
+type secretEventHandlerRegistration struct {
+	cache.ResourceEventHandlerRegistration
+	// objectKey will populate during AddEventHandler, and will be used during RemoveEventHandler
+	objectKey ObjectKey
+}
+
+func (r *secretEventHandlerRegistration) GetKey() ObjectKey {
+	return r.objectKey
+}
+
+func (r *secretEventHandlerRegistration) GetHandler() cache.ResourceEventHandlerRegistration {
+	return r.ResourceEventHandlerRegistration
+}
+
+type secretMonitor struct {
+	kubeClient kubernetes.Interface
+
+	lock sync.RWMutex
+	// monitors is map of singleItemMonitor. Each singleItemMonitor monitors/watches
+	// a secret through individual informer.
+	monitors map[ObjectKey]*singleItemMonitor
+}
+
+func NewSecretMonitor(kubeClient kubernetes.Interface) SecretMonitor {
+	return &secretMonitor{
+		kubeClient: kubeClient,
+		monitors:   map[ObjectKey]*singleItemMonitor{},
+	}
+}
+
+// create/update secret watch.
+// routeSecretName is a combination of "routename_secretname"
+func (s *secretMonitor) AddSecretEventHandler(namespace, routeSecretName string, handler cache.ResourceEventHandler) (SecretEventHandlerRegistration, error) {
+	return s.addSecretEventHandler(namespace, routeSecretName, handler, nil)
+}
+
+// addSecretEventHandler should only be used directly for tests. For production use AddSecretEventHandler().
+// createInformerFn helps in mocking sharedInformer for unit tests.
+func (s *secretMonitor) addSecretEventHandler(namespace, routeSecretName string, handler cache.ResourceEventHandler, createInformerFn func() cache.SharedInformer) (SecretEventHandlerRegistration, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if !validateName(routeSecretName) {
+		return nil, fmt.Errorf("invalid routeSecretName combination : %s", routeSecretName)
+	}
+
+	// Inside a namespace multiple route can refer a common secret.
+	// Hence route1_secret and route2_secret should denote separate key,
+	// so separate monitors inside a namespace.
+	key := NewObjectKey(namespace, routeSecretName)
+
+	// Check if monitor / watch already exists
+	m, exists := s.monitors[key]
+	if !exists {
+		secretName, err := getSecretName(routeSecretName, true)
+		if err != nil {
+			return nil, err
+		}
+
+		var sharedInformer cache.SharedInformer
+		if createInformerFn == nil {
+			// create a single secret monitor
+			sharedInformer = cache.NewSharedInformer(
+				cache.NewListWatchFromClient(
+					s.kubeClient.CoreV1().RESTClient(),
+					"secrets",
+					namespace,
+					fields.OneTermEqualSelector("metadata.name", secretName),
+				),
+				&corev1.Secret{},
+				0,
+			)
+		} else {
+			// only for testability
+			klog.Warning("creating informer for testability")
+			sharedInformer = createInformerFn()
+		}
+
+		m = newSingleItemMonitor(key, sharedInformer)
```

Comment:
the key for this monitor could be a generated UID or an incremented count.

Reply:
I think the `ObjectKey` with `namespace/secretName` should be sufficient as unique secret identifier. I'm not sure if we can fit in UID or an incremented count here.